src/datasources/cache/cache.first.data.source.ts

```diff
@@ -0,0 +1,64 @@
+import { NetworkRequest } from '../../common/network/entities/network.request.entity';
+import {
+  CacheService,
+  ICacheService,
+} from '../../common/cache/cache.service.interface';
+import {
+  INetworkService,
+  NetworkService,
+} from '../../common/network/network.service.interface';
+import { Inject, Injectable, Logger } from '@nestjs/common';
+import { LoggerMiddleware } from '../../common/middleware/logger.middleware';
+import { HttpErrorFactory } from '../errors/http-error-factory';
+
+/**
+ * A data source which tries to retrieve values from cache using
+ * {@link CacheService} and fallbacks to {@link NetworkService}
+ * if the cache entry expired or is not present.
+ *
+ * This is the recommended data source that should be used when
+ * a feature requires both networking and caching the respective
+ * responses.
+ */
+@Injectable()
+export class CacheFirstDataSource {
+  private readonly logger = new Logger(LoggerMiddleware.name);
+
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject(NetworkService) private readonly networkService: INetworkService,
+    private readonly httpErrorFactory: HttpErrorFactory,
+  ) {}
+
+  /**
+   * Gets the JSON payload behind {@link key}. If the value is not
+   * present tries to get the respective JSON payload from {@link url}
+   * Errors are not cached.
+   *
+   * @param key - the key to be used to retrieve from cache
+   * @param url - the HTTP endpoint to retrieve the JSON payload
+   * @param params - the parameters to be used for the HTTP request
+   * @param expireTimeSeconds - the time to live in seconds for the payload
+   * behind {@link key}
+   */
+  async get<T>(
```

Comment:
I know this is a first approach, but I'm tempted to comment that currently we're not using `params` at all regarding the cache (we're just using them for calling `NetworkService`), but probably we'll need to use those `params` to build the cache key in the future. Does it make sense to move the logic to compose the key to this class, in order to derive it from url and params, instead of making the caller responsible for sending the key? 

I mean, we could remove the `key` param from the signature of this method, and just deterministically derive the key inside this class.

We can think about this later and we can evaluate pros and cons (and feasibility) but it seems a good idea to decouple things and remove the responsibility from the caller of so it could even not know there's a cache in the next layer.

Reply:
> I know this is a first approach, but I'm tempted to comment that currently we're not using `params` at all regarding the cache (we're just using them for calling `NetworkService`), but probably we'll need to use those `params` to build the cache key in the future. Does it make sense to move the logic to compose the key to this class, in order to derive it from url and params, instead of making the caller responsible for sending the key?
> I mean, we could remove the `key` param from the signature of this method, and just deterministically derive the key inside this class.
> 
> We can think about this later and we can evaluate pros and cons (and feasibility) but it seems a good idea to decouple things and remove the responsibility from the caller of so it could even not know there's a cache in the next layer.

The idea that I had in mind with `CacheFirstDataSource` was that it would be just a layer in front of our cache that fallbacks to networking. I also wrote it in a way to be feature agnostic (from the cache perspective).

Features themselves currently do know about the cache via `CacheFirstDataSource` (ie.: it's not exposed as a "generic" `DataSource` which is something that we can also consider). I think it's an interesting alternative but by removing the cache key from the feature side we are tying the key to the parameters of HTTP.

I was considering that each feature could set a cache key that might independent from any URL parameter. If we remove the key from the feature side and rely on the URL (and additional params) as a key we need to make sure that the same keys are built when we also want to manually invalidate the cache. 🤔 

I agree that we should evaluate the next steps for this as any feature (even the current ones) will be based on this 👍

Reply:
Yes, invalidations were my main concern when suggesting this. Of course you have much more context than me and you know how we'll need to invalidate data. I'd say if invalidations are simple enough (data with a given key needs to be refreshed, so the key gets invalidated and that's it) we could push the key composition to the cache layer, but if invalidations could get complex (invalidate a collection of things, chaining invalidations, etc.) for me it's ok to let the services layer manage the keys and the logic. Let's discuss in the future, I'm happy with the current solution 👍